---
title: "Quantifying the uncertain evolutionary history of a cancer through clone trees"
author: | 
  | Lily Zheng, Laura Wood, Rachel Karchin, Rob Scharpf
  |
  |
  |
  | GI Spore Meeting
date: Feb. 10, 2020
output:
  beamer_presentation:
    theme: metropolis
    latex_engine: xelatex
---

# 

\includegraphics[width=1.1\textwidth]{ipmn_progression.png}

# 

\includegraphics[width=1.1\textwidth]{sample-tree.png}

# True clonal phylogeny

\includegraphics[width=1.1\textwidth]{clone-tree1.png}

- clone C is an early branch with sequence 5 as a private mutation

# Inferred sample tree 

\includegraphics[width=1.1\textwidth]{clone-tree2.png}

- the 5th sequence mutation appears as a parallel change between samples II and III


# 

\includegraphics[width=1.1\textwidth]{sample-tree-problem.png}

# Bayesian probabilistic inference of clone trees

Let $y_{is}$ denote the number of reads containing variant $i$ in sample $s$ and $Z_{i}$ the corresponding clone ($Z_{i}$ = 1, \ldots, K).

\begin{equation*}
  \begin{split}
    \left[y_{is} | \text{VAF}_{is}, n_{is} \right] &\sim \text{Binomial}(\text{VAF}_{is}, n_{is})\\
		[\text{VAF}_{is} | Z_{i}=z, m_{is}, c_{is}, \omega_{zs}] &= \cfrac{m_{is} \times \text{MCF}_{zs}}{c_{is} \times
                    \text{MCF}_{zs} + 2\times(1 - \text{MCF}_{zs})}\\
	\left[Z_{i} | \pi_1, \ldots, \pi_K, K \right] &\sim \text{Multinomial}(\pi_1, \ldots, \pi_{K})\\
    \text{MCF}_{zs} | G  &\sim \text{Beta}(a_G, b_G) 
  \end{split}
\end{equation*}

- Note the MCF is indexed by $z$ and is estimated from the set of all variants assigned to clone $z$

- The model induces a clustering of VAFs by the unobserved MCFs


# Given $K$, find $z$ and the MCF

- Variants in the same clone $z$ have the same unobserved MCF

- Neither $z$ or the MCF are observed

- Adopting a Bayesian hierarchical model for the observed VAF, we obtain a posterior distribution for the unobserved MCF and the clonal membership for each variant

  - As the counts are modeled directly, uncertainty reflected in the posteriors for these parameters reflects depth of coverage and the degree to which clones have distinguishable MCFs

# Simulated data with variants according to their assigned clone 

\includegraphics[width=\textwidth]{../output/plot_clusters.Rmd/simulated_vafs.pdf}

- the first clone is consistently high across all samples
- the third clone is present only in samples 1 and 3


# Same data

\includegraphics[width=\textwidth]{../output/plot_clusters.Rmd/simulated_vafs_not_grouped.pdf}

# Goodness of fit

\includegraphics[width=\textwidth]{../output/plot_clusters.Rmd/posterior_predictive.pdf}

# Posterior probability of clone assignments

\includegraphics[width=\textwidth]{../output/plot_clusters.Rmd/z_probs.pdf}
\vspace{-1em}
- What does this mean?

  - clone 1 is just a cluster of variants where the observed VAFs across the 3 samples can be thought of as having arised from just 3 MCFs


# MCFs

\includegraphics[width=\textwidth]{../output/plot_clusters.Rmd/CCF.pdf}

- Error bars are 95% posterior credible intervals

# Sampling Directed Acyclic Graphs (DAGs)

\includegraphics[width=\textwidth]{true_clone_tree.png}

- For 10 clones, there are $1.59 \times 10^{10}$ possible DAGs!

- Searching the space of all possible trees is computationally prohibitive


# Heuristic

\includegraphics[width=\textwidth]{crude_tree_structure2.png}

- $\approx$ 2 million DAGs compatible with these constraints

# Adjacency matrix of edges

\includegraphics[width=\textwidth]{adjacency_matrix.png}

- gray boxes indicate edges that are prohibited

# Conclusions and ongoing efforts 

- Inferential goals are modest:
  
  - to derive posterior probability distributions of the MCFs for each clone in each available sample
  
  - to derive posterior probability distributions for the DAGs depicting the evolutionary relationship among these clones 
   
   - the probability of a specific clone tree or a set of similar clone trees is easily obtained

- An efficient MCMC sampler for DAGs and application to experimental datasets are in progress












